tests/functional/test_fusion_pg.py

```diff
@@ -0,0 +1,1639 @@
+        tenant_name="tenant1",
+        tenant_space_name="tenant_space1",
+        placement_group_name="placement_group1",
+    )
+    pg_mock.delete_placement_group.assert_not_called()
+    op_mock.get_operation.assert_has_calls([call("op1"), call("op2")], any_order=True)
+
+
+@patch("fusion.OperationsApi")
+@patch("fusion.PlacementGroupsApi")
+@pytest.mark.parametrize(
+    ("raised_exception", "expected_exception"),
+    [
+        (purefusion.rest.ApiException, purefusion.rest.ApiException),
+        (HTTPError, HTTPError),
+    ],
+)
+def test_pg_create_triggers_update_exception(
+    pg_api_init, op_api_init, raised_exception, expected_exception
+):
+    module_args = {
+        "name": "placement_group1",
+        "display_name": "some_display_name",
+        "tenant": "tenant1",
+        "tenant_space": "tenant_space1",
+        "region": "region1",
+        "availability_zone": "availability_zone1",
+        "storage_service": "storage_service1",
+        "array": "array2",
+        "state": "present",
+        "issuer_id": "ABCD1234",
+        "private_key_file": "private-key.pem",
+    }
+    set_module_args(module_args)
+
+    get_placement_group_effects = [
+        None,
+        purefusion.PlacementGroup(
+            id="placement_group1_id",
+            name="placement_group1",
+            display_name="some_display_name",
+            self_link="test_self_link",
+            tenant=purefusion.TenantRef(
+                id="tenant1_id",
+                name="tenant1",
+                kind="Tenant",
+                self_link="some_self_link",
+            ),
+            tenant_space=purefusion.TenantSpaceRef(
+                id="tenant_space1_id",
+                name="tenant_space1",
+                kind="TenantSpace",
+                self_link="some_self_link",
+            ),
+            availability_zone=purefusion.AvailabilityZoneRef(
+                id="availability_zone1_id",
+                name="availability_zone1",
+                kind="AvailabilityZone",
+                self_link="some_self_link",
+            ),
+            placement_engine="heuristics",
+            protocols=[],
+            storage_service=purefusion.StorageServiceRef(
+                id="storage_service1_id",
+                name="storage_service",
+                kind="StorageService",
+                self_link="some_self_link",
+            ),
+            array=purefusion.ArrayRef(
+                id="array1_id", name="array1", kind="Array", self_link="some_self_link"
+            ),
+        ),
+    ]
+
+    def get_placement_group_effect(
+        tenant_name, tenant_space_name, placement_group_name
+    ):
+        i = get_placement_group_effects.pop(0)
+        if i is None:
+            raise purefusion.rest.ApiException()
+        return i
+
+    pg_mock = MagicMock()
+    pg_mock.get_placement_group = MagicMock(side_effect=get_placement_group_effect)
```

Comment:
You use this function several times and I believe it might be handy in some other tests too. What about putting it into the `utils.py`?

Something like:
```
def side_effect_popper_wrapper(side_effects):
     def side_effect_popper(*args, **kwargs):
             if len(side_effects) == 0:
                     raise purefusion.rest.ApiException()
             return side_effects.pop(0)
     return side_effect_popper
```

and in code

```suggestion
    pg_mock = MagicMock()
    pg_mock.get_placement_group = MagicMock(side_effect=side_effect_popper_wrapper(get_placement_group_effects))
```

And you can also remove `None` value from the list then.

Reply:
I think your logic is a bit different, your routine throws when the list runs out, mine when it encounters None. Nevertheless, factored out into utility function that returns from the list one by one and if it encounters an exception throws it instead